packages/pipeline/src/parsers/copper/index.ts

```diff
@@ -0,0 +1,274 @@
+import * as R from 'ramda';
+
+import { CopperActivity, CopperActivityType, CopperCustomField, CopperLead, CopperOpportunity } from '../../entities';
+
+const ONE_SECOND = 1000;
+export type CopperSearchResponse = CopperLeadResponse | CopperActivityResponse | CopperOpportunityResponse;
```

Comment:
Response types probably belong in __src/data_sources__ instead of here.

Reply:
I was thinking about this too. I originally put it in /data_sources, but there was a lot of referencing between files when writing the parsing logic. I think it makes development easier when the parser contains all of the logic/context of what a response looks like and how to transform it into an entity. Does that make sense?

Reply:
In general, types belong with the data that they describe. The data here is coming from the __data_sources/__ directory. In the current implementation, __data_sources/copper/index.ts__ contains the lines:

```
import {
    CopperActivityTypeCategory,
    CopperActivityTypeResponse,
    CopperCustomFieldResponse,
    CopperSearchResponse,
} from '../../parsers/copper';
```

You're importing types from another file in order to describe data that originates in this file, which seems wrong to me. You wouldn't make a __linked_list.ts__ file implementing a `LinkedList` and then put the types in a different file. I don't really see how this case is any different.

> but there was a lot of referencing between files when writing the parsing logic

I don't see how the current implementation reduces cross-file referencing in any significant way.  Could you elaborate?

Reply:
Yeah, I agree that conventionally data types should belong with the data where it originates. I'd argue that in this case, the logical responsibility for a) knowing what the data looks like and b) how to transform it both lie in the parser file. The data_sources file is only responsible for creating well-formed requests and managing pagination and other request logic. It doesn't have to know anything about the shape and types of data it's returning. In addition the practical benefit of having the typedef in the parser file where we are operating on it, is that we can write the parsing logic while having the data types right there in the same file to edit or reference. If we put the typedefs in the data_sources file, then we're cross-referencing /data_sources, /entities, and the /parser file. 

I do see the drawback of losing some breadcrumbs, e.g. someone who's only looking at the parser file can't easily follow the typedefs to the /data_sources file and see where the actual data is coming from. 

There's pros and cons each way -- happy to defer on this one if you think it's still better to put the typedefs in the /data_sources file.

Reply:
🤷‍♂️Not a hill I'm willing to die on so I guess just leave it.